bin/propolis-server/src/lib/migrate/mod.rs

```diff
@@ -378,106 +284,83 @@ pub(crate) async fn dest_initiate(
     ));
     info!(log, "Migration Destination");
 
-    let mut migrate_task = rqctx.context().objects.migrate_task.lock().await;
-
-    // This should be a fresh propolis-server
-    assert!(migrate_task.is_none());
-
-    // TODO(#165): https
-    // TODO: We need to make sure the src_addr is a valid target
-    let src_migrate_url =
-        format!("http://{}/instance/migrate/start", migrate_info.src_addr);
-    info!(log, "Begin migration"; "src_migrate_url" => &src_migrate_url);
-
-    let body = Body::from(
-        serde_json::to_string(&api::InstanceMigrateStartRequest {
-            migration_id,
+    let runtime_hdl = tokio::runtime::Handle::current();
+    tokio::runtime::Handle::current()
+        .spawn_blocking(move || -> Result<(), MigrateError> {
```

Comment:
Why does the dest side use `spawn_blocking` but not the source side?

Reply:
The big difference between the two sides is that the destination has to kick off and block on a new HTTP request within this closure, but the source side doesn't. `spawn_blocking` is this code path's way of contending with the async-ness of that request.

`request_migration_into` currently needs to issue the request and wait for the response synchronously to ensure that the controller state doesn't change while the connection is being established. The trouble here is that the "get an upgraded connection future" closure can't just use `block_on` by itself, because while `request_migration_into` isn't `async`, its caller is, so trying to call `block_on` panics at runtime with a "starting a runtime from within a task" error. Using `spawn_blocking` to create a top-level extra-runtime work item (that this function can then `await`) gets around that problem.

I experimented a bit with making `request_migration_into` async, such that it can just `await` the upgrade future, but that doesn't work (at least not the way I've tried it), because the controller lock is held for writing at the point where `request_migration_into` wants to make the request; trying to `await` that request while holding that lock causes a cascade of `future returned by $FUNCTION is not Send` errors.

With all that said, it might be possible to pry this apart by making `request_migration_into`'s `upgrade_fn` into a future that outputs a `Result<OnUpgrade, MigrateError>` and stashing that in the `ExternalRequest::MigrateAsTarget` message. That also gets around the recursive-runtime problem by pushing everything to the state worker, which is already a free thread. I can give this a whirl but might be inclined to file an issue and try this in a follow-on change.

Reply:
If it's not going to immediately blow up in our faces, a follow-up would be preferred.  I suspect we'll get a better feel for how we want to continue shaping these bits as we plug in more pieces.